### PR TITLE
[GEOS-11379] Refactor inline JavaScript in the OGC API modules

### PR DESCRIPTION
**Pull request by sikeoka**
[GEOS-11379](https://osgeo-org.atlassian.net/browse/GEOS-11379) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

This PR moves the inline JavaScript in all of the OGC API community modules into external files. This PR is related to Content-Security-Policy work for GeoServer 2.26.0 and should **NOT** be backported.

All Java inputs that were previously written directly into the script tag are now written to hidden input fields that the external JavaScript file will read. Enabling HTML escaping was necessary to not introduce new vulnerabilities with this PR but that also required changing how the pagecrumbs are initialized everywhere.

This PR uses the new webresources path added to gs-main by https://github.com/geoserver/geoserver/pull/7554 which does not affect unit tests but is required if attempting to manually run GeoServer with these changes.
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by jodygarnett** [comment]
Thanks for being patient on this one @sikeoka.

GeoCat offered to test this a couple meetings ago - and @MichelGabriel is now available:

- He has experience building and testing the OGCAPI module
- Better yet he has the JavaScript experience to appreciate the change (and troubleshoot if necessary)

@MichelGabriel please review this PR:
- The timeframe is prior to the September 2.26 release cycle
- I have sent you an invite to "Team GeoServer Wiki" - which provide enough permissions that I can assign this PR to you for review

**Comment by davidblasby** [comment]
I am taking a quick look at this.

First thing I notice is that the deployed ogc-api are NOT listed on the Server-Status/Modules page.  Nothing to do with this PR, but it might be nice to see it listed there.

**Comment by davidblasby** [comment]
ok, I havent been able to run this version of geoserver.  I get an error like this;

```
Caused by: java.lang.NoSuchFieldError: QUETTA
    at org.geotools.measure.PrefixDefinitions.<clinit> (PrefixDefinitions.java:52)
    at org.geotools.measure.UnitDefinition.withStandardPrefixes (UnitDefinition.java:47)
    at org.geotools.measure.UnitDefinitions.<clinit> (UnitDefinitions.java:54)
    at org.geotools.measure.WktUnitFormat.<clinit> (WktUnitFormat.java:45)
    at org.geotools.measure.Units.autoCorrect (Units.java:159)
```

Looking over at Geotools, I see this commit (about 2 months ago):

https://github.com/geotools/geotools/commit/29688d999077df46b29228dac7f6271a1b12a8b1

Which is probably the root problem.  I expect there is some sort of GT version miss-match.

**Comment by davidblasby** [comment]
I tried building GT locally, and havent figure out how to run this PR.

File: src/community/ogcapi/ogcapi-tiles/src/main/resources/org/geoserver/ogcapi/v1/tiles/tileMatrixSet.ftl [hunk 1]
```diff
@@ -1,5 +1,9 @@
 <#global pagetitle=model.id>
-<#global pagecrumbs="<li class='breadcrumb-item'><a href='"+serviceLink("")+"'>Home</a></li><li class='breadcrumb-item'><a href='"+serviceLink("tileMatrixSets")+"'>Tile Matrix Sets</a></li><li class='breadcrumb-item active'>"+pagetitle+"</li>">
+<#global pagecrumbs>
+  <li class='breadcrumb-item'><a href='${serviceLink("")}'>Home</a></li>
+  <li class='breadcrumb-item'><a href='${serviceLink("tileMatrixSets")}'>Tile Matrix Sets</a></li>
+  x<li class='breadcrumb-item active'>${pagetitle}</li>
```
Comment (davidblasby) [inline src/community/ogcapi/ogcapi-tiles/src/main/resources/org/geoserver/ogcapi/v1/tiles/tileMatrixSet.ftl:5]: This line (#5) starts with an "x" - this shows in the HTML output.

**Comment by davidblasby** [comment]
I noticed that http://localhost:8888/geoserver/collections/tiger:poly_landmarks/queryables?f=application%2Fjson
gives a 404 (HTML page is ok).
However, this happens in the old and new version (i.e. not caused by this PR).

**Comment by davidblasby** [comment]
I noticed that the dropdown for downloading the data in another format looks like this:

`<option value="http://localhost:8080/geoserver/ogc/features/v1/collections/tiger:poly_landmarks/items?f=text%2Fcsv">text/csv</option>`

However the actual request is `http://localhost:8080/geoserver/ogc/features/v1/collections/tiger:poly_landmarks/items?f=text%2Fcsv&limit=50` (with a &limit=50).  This was unexpected.

I think this is the old behaviour as well, but it took me quite a while to track down what was happening (dowload links are very hard to trace in Chrome DevTools).

**Comment by davidblasby** [comment]
Also noticed that most of the conformance links (from http://localhost:8080/geoserver/ogc/features/v1/conformance?f=text%2Fhtml) don't go anywhere.  Only first few do - the rest return a "not found".

Example:
https://geoserver.org/spec/ecql/1.0/req/ecql-text
http://www.opengis.net/spec/cql2/1.0/req/spatial-operators

(only first few links work)

```
http://www.opengis.net/spec/ogcapi-features-1/1.0/conf/core
http://www.opengis.net/spec/ogcapi-features-1/1.0/conf/oas30
http://www.opengis.net/spec/ogcapi-features-1/1.0/conf/html
http://www.opengis.net/spec/ogcapi-features-1/1.0/conf/geojson
http://www.opengis.net/spec/ogcapi-features-2/1.0/conf/crs
http://www.opengis.net/spec/ogcapi-features-3/1.0/req/features-filter
http://www.opengis.net/spec/ogcapi-features-3/1.0/req/filter
http://www.opengis.net/spec/ogcapi-features-5/1.0/req/search
http://geoserver.org/spec/ecql/1.0/req/gs-ecql
http://geoserver.org/spec/ecql/1.0/req/ecql-text
http://www.opengis.net/spec/cql2/1.0/req/basic-cql2
http://www.opengis.net/spec/cql2/1.0/req/advanced-comparison-operators
http://www.opengis.net/spec/cql2/1.0/req/arithmetic
http://www.opengis.net/spec/cql2/1.0/req/property-property
http://www.opengis.net/spec/cql2/1.0/req/basic-spatial-operators
http://www.opengis.net/spec/cql2/1.0/req/spatial-operators
http://www.opengis.net/spec/cql2/1.0/req/functions
http://www.opengis.net/spec/cql2/1.0/req/cql2-text
http://www.opengis.net/spec/ogcapi-records-1/1.0/req/sorting
http://www.opengis.net/spec/ogcapi-features-1/1.1/req/ids
http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/core
http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/collections
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/core
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/tileset
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/multitiles
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/info
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/tmxs
http://www.opengis.net/spec/tilematrixset/1.0/conf/tilematrixset
http://www.opengis.net/spec/tilematrixset/1.0/conf/json-tilematrixset
```

**Comment by davidblasby** [comment]
Moving on to OGCAPI-maps.  

1. The API Definition doesn't go anywhere (both old and new):
http://localhost:8080/geoserver/ogc/maps/v1/openapi?f=text%2Fhtml 

2. All the conformance links go to "not found" (except the 1st one):
http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/core
http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/collections
http://www.opengis.net/spec/ogcapi-maps-1/1.0/conf/core
http://www.opengis.net/spec/ogcapi-maps-1/1.0/conf/geodata
http://www.opengis.net/spec/ogcapi-maps-2/1.0/conf/bbox
http://www.opengis.net/spec/ogcapi-maps-2/1.0/conf/crs


(none of these are new issues)

**Comment by davidblasby** [comment]
For OGCAPI Tiles:

Some conformance links don't go anywhere ("Not found"):
http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/collections
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/multitiles
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/info
http://www.opengis.net/spec/ogcapi-tiles-1/1.0/conf/tmxs
http://www.opengis.net/spec/tilematrixset/1.0/conf/tilematrixset
http://www.opengis.net/spec/tilematrixset/1.0/conf/json-tilematrixset


(this isn't a new issue)

**Comment by davidblasby** [comment]
For OGCAPI-Coverages:

1. Most of the conformance links go to "not found":
http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/collections
http://www.opengis.net/spec/ogcapi-common/1.0/req/oas30
http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/geodata
http://www.opengis.net/spec/ogcapi-coverages-1/1.0/conf/geodata-coverage
http://www.opengis.net/spec/ogcapi-coverages-1/1.0/conf/geotiff
http://www.opengis.net/spec/ogcapi-coverages-1/1.0/req/coverage-subset

2. Several of the standard `release` coverages wouldn't render:
{"type":"NoApplicableCode","title":"Cannot encode this coverage in the requested format\n- Unable to render RenderedOp for this operation."}
{"type":"NoApplicableCode","title":"Could not find acceptable representation"}

3. some of the coverage formats wouldn't produce results like `text/html`

(these are issues with both old and new)

**Comment by davidblasby** [comment]
OGCAPI-Styles:

1. ALL Conformance links go to "not found":
http://www.opengis.net/t15/opf-styles-1/1.0/conf/core
http://www.opengis.net/t15/opf-styles-1/1.0/conf/html
http://www.opengis.net/t15/opf-styles-1/1.0/conf/json
http://www.opengis.net/t15/opf-styles-1/1.0/conf/mapbox-styles
http://www.opengis.net/t15/opf-styles-1/1.0/conf/sld-10
http://www.opengis.net/t15/opf-styles-1/1.0/conf/sld-11

2. The ".properties" (converted). `Stylesheet as Property 1.0.0 (converted)` doesn't seem to correctly do the conversion (just gives a file with one line: `type=line`

(these are issues with both old and new)

**Comment by davidblasby** [comment]
For DDGS:

The conformance links (only one) goes to "Not Found".
http://www.opengis.net/spec/ogcapi-dggs-1/1.0/conf/core

(issue with both old and new)

**Comment by davidblasby** [comment]
OGCAPI Images:

1. Several of the conformance links go to "not found":

http://www.opengis.net/spec/ogcapi-common-2/1.0/conf/collections
http://www.opengis.net/spec/ogcapi-images-1/1.0/req/core
http://www.opengis.net/spec/ogcapi-images-1/1.0/req/transactional

2. I notice the breadcrumbs are NOW correct:
image

3. The `Images available for this collection.` doesn't seem to go anywhere.

(these are issues with both old and new)

**Comment by sikeoka** [comment]
There are so many comments here unrelated to the actual changes of the PR that I can't be sure what I might need to fix.

**Comment by davidblasby** [comment]
@sikeoka  - Hi, Steve.  Yes, I did make a bunch of notes.  I was pretty explicit in things that were in both the old and new - `(these are issues with both old and new)`.  However, since I was doing A-B testing and clicking on everything, I just noted issues I found.  I can put them in a ticket - sorry for the clutter.

1. I wasn't able to find any javascript on any of the pages -- good job.
2. I had to do a `git merge origin/main` because of recent GT changes that likely have a recent GS component (cf. 'QUETTA', above).  I'd recommend either merging or rebasing...
3. As I pointed out (with pic and line #), there's an "x" in the breadcrumb output for the tilematrix.  Trivial fix.

Other than that, I did find some issues that should be addressed, but they ALL are in both the old and new version - so not your problem.  I'll see if I can put in a separate ticket.

Good job!

**Comment by sikeoka** [comment]
I fixed the extra "x" and all of the PR checks so a rebase isn't necessary.

**Comment by davidblasby** [comment]
> I fixed the extra "x" and all of the PR checks so a rebase isn't necessary.

yes - the checks pass, but it will not run as-is.  you'll get the `QUETTA` problems.

**Comment by sikeoka** [comment]
I rebased the branch and everything looks good.

**Comment by sikeoka** [comment]
Is there anything else that needs to be addressed for this PR?

**Comment by jodygarnett** [comment]
I think we want confirmation with David that each screen has been checked and then we are good to go?

@aaime did you need anything else?

**Comment by aaime** [comment]
Like Steve, I was surprised by David's feedback, he checked stuff that was not needed... and he's welcomed to fix every one of those issues in a separate PR.

Indeed all that was needed was pure A-B testing: is the current page working and looking like it used to, or not?
I don't see such information presented here.

David, even if you were "just taking notes", there are many suitable places to add them, a local notepad, a wiki page, a but report... this PR never was a good one.

**Comment by jodygarnett** [comment]
@groldan you were looking at what is needed for OGCAPI-Features? Perhaps you can collect David’s testing notes above.

Reading the above I am trying to spot any outstanding differences due to this PR?

I am unclear if this is a change due to the PR or just something that surprised David:

> I noticed that the dropdown for downloading the data in another format looks like this:
> 
> `<option value="http://localhost:8080/geoserver/ogc/features/v1/collections/tiger:poly_landmarks/items?f=text%2Fcsv">text/csv</option>`
> 
> However the actual request is `http://localhost:8080/geoserver/ogc/features/v1/collections/tiger:poly_landmarks/items?f=text%2Fcsv&limit=50` (with a &limit=50). This was unexpected.
> 
> I think this is the old behaviour as well, but it took me quite a while to track down what was happening (dowload links are very hard to trace in Chrome DevTools).

**Comment by sikeoka** [comment]
That is not something that was changed for this PR. The old behavior was to have the URLs without the limit in the `select` list and when the user selects a format, JavaScript will append the limit to the URL and then open the modified URL. This PR preservers that behavior and just moves it out of inline JavaScript. The limit could be appended to the URL in the FreeMarker template and then the JS would just open the URL but that specific refactoring was unnecessary for this PR.

**Comment by davidblasby** [comment]
I didn't notice anything in my A/B testing.

This is approved for merging.

**Review by davidblasby — APPROVED** [review]
No differences found in A/B testing.

**Pull request merged**